### ISSUE TITLE
Support for Azure Service Fabric

### ISSUE THREAD

**Issue by MedAnd** [dapr/dapr#630]
As Dapr is being envisioned as a modern, cloud native, event-driven, cross platform runtime for microservices, please add support and integration with Azure Service Fabric.

/area runtime
/area docs

- Support for deploying & running the Dapr runtime as a system service.
- Integration of Dapr and Service Fabric Actors, Reliable Collections etc
- Documentation and integration samples for stateful and stateless scenarios & extension and plugin models etc
- Developer experience such as tooling, debugging, deployment etc

**Comment by brendandburns** [comment]
Since Dapr is a collection of Linux containers, there should be no problems running dapr alongside Service Fabric using Docker on Linux or [Linux Containers on Windows](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/linux-containers)

We'd welcome contribution/PRs to make this work, absent that, the work is not currently on the dapr roadmap.

**Comment by MedAnd** [comment]
@brendandburns - appreciate the engagement and transparency! Can we read above as confirmation of LCOW support on Service Fabric (announcement pending Ignite)?

**Comment by aL3891** [comment]
Hang on, you're saying you only plan to have dapr running in linux containers on windows? Windows it self is not a primary runtime target?

**Comment by MedAnd** [comment]
Native Service Fabric and Dapr integration would be ideal... I was thinking why not run Dapr as a SF System Service, similar to the Managed Identity system service which is in preview... also why pay the overhead of containers when not required? 

@aL3891 - LCOW support in Service Fabric is at least a new feature ;-) but hope @brendandburns will confirm.

**Comment by msfussell** [comment]
Dapr can run as a side-car process to the app, it does not need to be containerized. At the very least right now you can create a service fabric stateless service with two code packages, one for your code and the other using the dapr.exe of your choice (windows/linux etc) placing the component file in a /component folder in the package. The deploy this into your service fabric cluster. This will enable you to use some of the functionality such as bindings and external state stores such as CosmosDB. 

Native Service Fabric integration is needs for the other capabilities. For example to get Dapr service invocation this can integrate with the SF reverse proxy on each machine. This would require creating a service that mapped Dapr app id to SF app id to build the SF URL format https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy. Starting a design doc for these native SF services, it what we can do on this issue.

**Comment by mkosieradzki** [comment]
How about using SF Service Discovery API directly and caching the result in Dapr?

**Comment by mkosieradzki** [comment]
BTW. @msfussell any idea how to do code-package auto-injection?

**Comment by amanbha** [comment]
> BTW. @msfussell any idea how to do code-package auto-injection?

This needs to be done manually in manifests, SF doesn't allow auto injection.

**Comment by mkosieradzki** [comment]
@amanbha thanks - so we should go with a CI/CD step, patching the manifest. However manual should work as well.
Also thinking about configuration delivery - standalone config delivered via a config/data package should be more Service Fabric-ish than a central operator like in k8s version.

Even if managed via a central "operator" it should be deployed and upgrade as an SF config/data package imo. Does it makes sense?

**Comment by mkosieradzki** [comment]
https://github.com/dapr/dapr/blob/master/docs/decision_records/architecture/ARC-001-refactor-for-modularity-and-testability.md - we would require a pluggable service discovery support.

**Comment by mkosieradzki** [comment]
@amanbha what do you think would be better:
1. go with SF Reliable Actors directly?
2. deploy the placement service to the SF?

Option one: sounds more native and reliable
Option two: sounds easier to maintain and future proof in case placement service will be responsible for something more than actors.

Is there any compatibility layer in the SDK to use SF actors and Dapr actors interchangeably?

**Comment by amanbha** [comment]
> @amanbha what do you think would be better:
> 
> 1. go with SF Reliable Actors directly?
> 2. deploy the placement service to the SF?
> 
> Option one: sounds slightly more native and reliable.
> Option two: sounds easier to maintain and future proof in case placement service will be responsible for something more than actors.
> 
> Is there any compatibility layer in the SDK to use SF actors and Dapr actors interchangeably? I am also concerned about the requests coming from other services.
> 
> I am leaning towards the option 2, but maybe I am missing something?

If you are on SF, go with SF Reliable Actors as it provides none, volatile and persisted state stores and state stored within the cluster. SF Actors are battle tested for few years now with real high scale prod workloads running on it.
Dapr is an alpha release and we want it to reach maturity and scale soon,once it reaches GA, then it will be more future proof to use as it allows calls over http from other languages as well.
There is no compatibility layer to call one or the other and that's not the goal of Dapr SDKs. The apis in dapr are however very similar to SF Actors, so that atleast the service side code for SF Actors can be made to run on Dapr with little changes, then client calls can come in via remoting or over http.

**Comment by amanbha** [comment]
> https://github.com/dapr/dapr/blob/master/docs/decision_records/architecture/ARC-001-refactor-for-modularity-and-testability.md - we would require a pluggable service discovery support.

Yes, it should be pluggable, could you please an issue for it.

**Comment by amanbha** [comment]
> @amanbha thanks - so we should go with a CI/CD step patching the manifest.
> However manual should work as well.
> 
> Also thinking about configuration delivery - standalone config delivered via a config/data package should be more Service Fabric-ish than a central operator like in k8s version.

Yes, it could be in CI CD pipeline. The config approach suits better for SF, but it's rolling upgrade so instances would be running with different configs until the upgrade finishes.

**Comment by mkosieradzki** [comment]
> The config approach suits better for SF, but it's rolling upgrade so instances would be running with different configs until the upgrade finishes.

But this is by design right? :) This is exactly how it should work on SF, which is about the resilliency.

**Comment by mkosieradzki** [comment]
@amanbha btw. do you think that side-by-side code packages is the optimal design? Maybe a co-located stateful service is the answer? :)

**Comment by amanbha** [comment]
> > The config approach suits better for SF, but it's rolling upgrade so instances would be running with different configs until the upgrade finishes.
> 
> But this is by design right? :) This is exactly how it should work on SF, which is about the resilliency.

Yes, that's by design. I was thinking long the lines that you could change components.yaml for Dapr to save state into cosmos instead of redid and that wouldn't go well while rolling upgrade is going on, you might want the switch to be instantaneous across a instances rather than rolling.

**Comment by amanbha** [comment]
> @amanbha btw. do you think that side-by-side code packages is the optimal design? Maybe a co-located stateful service is the answer? :)

It needs to be tried out to see what works better. Dapr needs to know app port and app needs to know dapr port,so colocating them on a node with static port on a node, then providing the required info to dapr  via cmdline and to app via envvar  would work.

**Comment by heavenwing** [comment]
@msfussell how to run dapr with "dapr run" in below situation?

> Dapr can run as a side-car process to the app, it does not need to be containerized. At the very least right now you can create a service fabric stateless service with two code packages, one for your code and the other using the dapr.exe of your choice (windows/linux etc) placing the component file in a /component folder in the package. The deploy this into your service fabric cluster. This will enable you to use some of the functionality such as bindings and external state stores such as CosmosDB.

**Comment by mkosieradzki** [comment]
> @msfussell how to run dapr with "dapr run" in below situation?

`daprd --help` should show you all options (most of default looks very reasonable for SF).

At least according to the output of `daprd --help` component file should be placed in `components` not `component` folder.

**Comment by mkosieradzki** [comment]
If you want to have actors running you should also run placement service available here as singleton and you can add: `-placement-address` to your daprd commandline. Placement service does not use any underlying storage (it just brodcasts the DHT updates). It can be downloaded from here as well: https://github.com/dapr/dapr/releases

**Comment by amanbha** [comment]
> If you want to have actors running you should also run placement service available here as singleton and you can add: `-placement-address` to your daprd commandline. Placement service does not use any underlying storage (it just brodcasts the DHT updates). It can be downloaded from here as well: https://github.com/dapr/dapr/releases

It would be little tricky in SF, the resolved address needs to be assed to reresolve when placement service moves, maybe it could use dns name. It needs to be exposed to both actor services and to services which acts as clients.

**Comment by LMWF** [comment]
Here is an sample showing how to run the local Dapr Hello World sample on SF, to show basic functionality such as using a state store.
https://github.com/LMWF/samples/blob/sfsample/1.hello-world/servicefabricsample.md

**Comment by tg123** [comment]
@LMWF 
Seems it is a guide converting any executable, dapr in your sample, into SF application format and run it on SF.

Is there any plan to build a native support on SF
e.g.
`dapr run` -> submit app to SF environment ?

**Comment by amanbha** [comment]
@tg123 dapr run is cli command for local development. The run cli command will not submit app to any  platform. The deployments still need to be platform specific.

**Comment by sandeep-ma** [comment]
> Since Dapr is a collection of Linux containers, there should be no problems running dapr alongside Service Fabric using Docker on Linux or [Linux Containers on Windows](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/linux-containers)
> 
> We'd welcome contribution/PRs to make this work, absent that, the work is not currently on the dapr roadmap.

@brendandburns , that is a silly idea. Service Fabric is closed source and the equivalent APIs already exist there. More interesting would be to have dapr available as part of a serverless Mesh offering, as an alternative to Kuberneties and Helm, for Windows Containers on Azure Container Instances. It would be interesting to know what is Chacko's view point... or is something already in the works?

**Comment by yaron2** [comment]
@sandeep-ma we'd like to refrain from calling other people's comments "silly", especially if they took the time to answer someone else's question.

The idea of having Dapr automatically injected and managed in serverless environments is interesting, and has been brought several times.

**Comment by MedAnd** [comment]
From @markrussinovich's talk he mentions Dapr was designed to run both as a container and process side car. It seems for Service Fabric integration at least, both container and process daprization options are possible but would argue having process (aka native) level Service Fabric integration should be the first option as it's better for performance, resource utilization etc? Interested in what @msfussell and team think.

**Comment by msfussell** [comment]
Leon's sample here https://github.com/LMWF/samples/blob/sfsample/1.hello-world/servicefabricsample.md shows simply how Dapr can be partially used today deployed as a process. And with this https://github.com/dapr/dapr/issues/753 we can now get Service Fabric integrated. However there is not an finalize SF design yet. I would fully expect these to run as process through, rather than containers in SF.

**Comment by abatishchev** [comment]
It's hard to think more pointless than a bot that closed issues because they're "stale". Remove it now.

**Comment by acartcat** [comment]
Given that SF provides relatively :-) straight forward path to process orchestration I would expect there would be many scenarios where even partial Dapr support would be beneficial for application development.  Certainly in our case which is largely guest executables.  At a minimum a section in the documentation for the SF community, what works at the start, what can work with a bit of effort.  Maybe I've just described what I can do as a contribution but at least I can keep this issue going towards a suitable conclusion.

**Comment by sandeep-ma** [comment]
Good idea... but do sf and mesh still exist?

**Comment by msfussell** [comment]
A sample has been provided here https://github.com/dapr/samples/tree/master/hello-service-fabric
Service Fabric is still very much in existence and extensively used. SF Mesh evolved into Container Apps as a more general version and in essence very much showed the path to the serverless container offering.

**Comment by KaiWalter** [comment]
The guest EXE model is fine for a start and a very basic deployment. Question for me is, whether it makes sense or whether there is any priority that

a) Dapr works with Service Fabric name resolution at some time to allow service resolution across multiple apps
b) Service Fabric will have support for sidecars or something like pods at any time or whether we only have this guest EXE stacking inside one container as an option

Anyway this is no priority for us anymore - as we will shift the workloads in question to Container Apps.

**Comment by acartcat** [comment]
Kai,
Thanks to Mark back in 2016 the company I work for has a global solution
using Service Fabric and relies on guest executables as the primary
application deployment pattern.  Given that SF is a process orchestrator in
addition to a container orchestrator we have great flexibility for building
apps, in our case Java/Spring.  We use a combination of SF DNS service
along with the reverse proxy. I will be spending more time on exploring how
we can integrate dapr into our existing and future work.

**Comment by kchudnovskiy** [comment]
@msfussell , have there been any progress on service invocation? The way I am reading your earlier comment we would need native support and not be able to make it work ourselves. 

Native Service Fabric integration is needs for the other capabilities. For example to get Dapr service invocation this can integrate with the SF reverse proxy on each machine. This would require creating a service that mapped Dapr app id to SF app id to build the SF URL format https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy. Starting a design doc for these native SF services, it what we can do on this issue.
